### PR TITLE
Add dynamic ACK support.

### PR DESCRIPTION
**Pull request by boaks**
Use piggybacked responses, if amqp response is fast available.

Signed-off-by: Achim Kraus <achim.kraus@bosch.io>

### CODE REVIEW

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java [hunk 1]
```diff
@@ -1059,6 +1066,36 @@ private void addCommandReceptionTimer(
...
     }
 
+    /**
+     * Sets a timer to trigger the sending of a separate ACK to a device.
+     * <p>
+     *
+     * @param context The device's currently executing HTTP request.
+     * @param timeoutMillis The number of milliseconds to wait for a separate ACK.
+     */
+    private void addSeparateAcceptTimer(
```
Comment (sophokles73) [inline adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java:1076]: how about moving this method to CoapContext?

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java [hunk 2]
```diff
@@ -352,4 +357,39 @@ public final void setMessageOffloadingEnabled(final boolean messageOffloading) {
         this.messageOffloadingEnabled = messageOffloading;
     }
 
+    /**
+     * Gets the timeout to ACK a CoAP CON requests.
+     * <p>
+     * If the response is available before that timeout, a more efficient piggybacked response is used. If the timeout
+     * is reached without response, a separate ACK is sent to prevent the client form retransmit the CON request.
+     * <p>
+     * The default value of this property is {@link #DEFAULT_TIMEOUT_TO_ACK}, {@value #DEFAULT_TIMEOUT_TO_ACK}
```
Comment (sophokles73) [inline adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java:366]: IMHO the `{@link }` should be omitted

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java [hunk 3]
```diff
@@ -352,4 +357,39 @@ public final void setMessageOffloadingEnabled(final boolean messageOffloading) {
...
+        return timeoutToAck;
+    }
+
+    /**
+     * Sets the timeout to ACK a CoAP CON request.
+     * <p>
+     * If the response is available before that timeout, a more efficient piggybacked response is used. If the timeout
+     * is reached without response, a separate ACK is sent to prevent the client form retransmit the CON request.
+     * <p>
+     * The default value of this property is {@link #DEFAULT_TIMEOUT_TO_ACK}, {@value #DEFAULT_TIMEOUT_TO_ACK}
```
Comment (sophokles73) [inline adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java:382]: see above

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java (hunk 1 above)
Comment (boaks) [reply adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java]: Done.

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java (hunk 2 above)
Comment (boaks) [reply adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java]: Done.

File: adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java (hunk 3 above)
Comment (boaks) [reply adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapAdapterProperties.java]: Done.

**Pull request merged**